File: tools/spikes/printedcard/card.mjs
// Renders a printed card from a real issued credential, and decodes it back.
//
// #66's second leg. PixelPass is the format Blueprint §5 names for the printed
// path: the QR carries the *full signed VC*, so a bare scan verifies signature
// and schema with no network at all. That is what makes a worker's record
// provable to a stranger without CREST, and it is the reason the payload is not
// a URL or a lookup key — either of those would put a server between a worker
// and their own history.
//
// Usage: node card.mjs <credential.json> <outdir>
import { readFileSync, writeFileSync } from 'node:fs'
import { join } from 'node:path'
import pkg from '@mosip/pixelpass'

const { generateQRData, generateQRCode, decode } = pkg

const [, , credPath, outDir] = process.argv
if (!credPath || !outDir) {
  console.error('usage: node card.mjs <credential.json> <outdir>')
  process.exit(2)
}

const credential = JSON.parse(readFileSync(credPath, 'utf8'))
const flat = JSON.stringify(credential)

// PixelPass compresses (zlib) and base45-encodes. No header is passed: an
// empty header is what Inji's own wallet writes, and a card only this repo can
// read would prove nothing about interoperability.
const payload = generateQRData(flat)
writeFileSync(join(outDir, 'card.qr.txt'), payload)

const png = await generateQRCode(flat)
writeFileSync(join(outDir, 'card.png'), Buffer.from(png.split(',')[1], 'base64'))

// The round trip is the claim worth checking here: a card that renders but
// decodes to something else is a card that fails in a worker's hands, in a
// place where nobody can debug it.
const back = decode(payload)
if (back !== flat) {
  console.error('[FAIL] the card does not decode back to the credential that was printed')
  process.exit(1)
}
writeFileSync(join(outDir, 'decoded.json'), back)

const subject = credential.credentialSubject ?? {}
const html = `<!doctype html><meta charset="utf-8"><title>CREST work record</title>
<style>
 body{font:14px/1.5 system-ui,sans-serif;margin:0;display:flex;justify-content:center;padding:24px}
 .card{width:340px;border:1px solid #222;border-radius:10px;padding:18px}
 h1{font-size:15px;margin:0 0 2px} .sub{color:#555;font-size:12px;margin:0 0 14px}
 img{width:100%;image-rendering:pixelated;border:1px solid #eee}
 dl{display:grid;grid-template-columns:auto 1fr;gap:2px 10px;font-size:12px;margin:12px 0 0}
 dt{color:#555} dd{margin:0}
 .foot{margin-top:12px;font-size:10px;color:#666;border-top:1px solid #eee;padding-top:8px}
 @media print{body{padding:0}}
</style>
<div class="card">
  <h1>${escapeHtml(String(subject.activity ?? 'work record'))}</h1>
  <p class="sub">${escapeHtml(String(subject.issuerOrg ?? ''))}</p>
  <img alt="PixelPass QR carrying the full signed credential" src="data:image/png;base64,${png.split(',')[1]}">
  <dl>
    <dt>outcome</dt><dd>${escapeHtml(String(subject.outcome?.value ?? ''))} ${escapeHtml(String(subject.outcome?.unit ?? ''))}</dd>
    <dt>period</dt><dd>${escapeHtml(String(subject.period?.start ?? ''))}</dd>
    <dt>definition</dt><dd>${escapeHtml(String(subject.definition?.ref ?? ''))} v${escapeHtml(String(subject.definition?.version ?? ''))}</dd>
  </dl>
  <p class="foot">Scanning this code reads the whole signed record. It carries no trust tier and no
  identity number: strength is worked out by whoever checks it, from the facts inside.</p>
</div>`
writeFileSync(join(outDir, 'card.html'), html)

function escapeHtml (s) {
  return s.replace(/[&<>"]/g, c => ({ '&': '&amp;', '<': '&lt;', '>': '&gt;', '"': '&quot;' }[c]))
}

console.log(`[PASS] card rendered and decodes back byte-identical (${payload.length} chars of QR payload)`)

File: services/core/verification/card.go
package verification

import (
	"encoding/base64"
	"encoding/json"
	"fmt"
	"html"
	"net/http"
	"strings"
	"time"

	"rsc.io/qr"

	"github.com/theflywheel/crest/pkg/credential"
	"github.com/theflywheel/crest/pkg/httpx"
	"github.com/theflywheel/crest/pkg/store"
)

// The printed card (#24, Blueprint §5).
//
// One of three ways a worker holds a credential, and the only one available to
// a worker without a phone — which makes it the holding mechanism for exactly
// the people the inclusion path exists for. Not a fallback: for them it is the
// wallet.
//
// Printed at issuance rather than at enrolment, settled 2026-08-24 on #24. At
// enrolment no work has been done and there is no credential to print; §5 used
// to say "from enrolment", which read as though a worker walked away from
// registration already holding their history. The card is produced through the
// same assisted-access channel enrolment uses.
//
// The QR carries the whole signed credential rather than a link to one. That is
// the property the offline path exists for: a bare scan verifies signature and
// schema with no network, so the record is provable to a stranger who has never
// heard of CREST and has no signal. A QR holding a URL would move the proof back
// onto a server and the worker back into needing one.
//
// What it deliberately does not carry: any trust tier. Strength is derived by
// whoever checks it, from the provenance facts inside — printing a tier would
// freeze a judgement onto a piece of paper that outlives the reasons for it.

// cardQRLevel is the error-correction level.
//
// Medium, not Low. A printed card lives in a pocket for months, gets rained on
// and folded, and the failure this guards against is a worker holding a card
// nobody can scan — which is indistinguishable, to them, from having no record
// at all. Higher levels than this would make the QR denser than a phone camera
// reliably reads at arm's length, which is the same failure from the other side.
const cardQRLevel = qr.M

func (h *handlers) credentialCard(w http.ResponseWriter, r *http.Request) {
	c, err := getCredential(r.Context(), h.d.DB.Q(), r.PathValue("id"))
	if err != nil {
		httpx.NotFoundOr(w, h.d.Log, "credential", err, store.ErrNotFound)
		return
	}

	payload, err := credential.EncodePixelPass(c.Doc)
	if err != nil {
		httpx.Fail(w, h.d.Log, "encode the card payload", err)
		return
	}

	// The payload alone, for a caller that renders its own card — a print
	// station with its own stationery, or a test.
	if r.URL.Query().Get("format") == "payload" {
		w.Header().Set("Content-Type", "text/plain; charset=utf-8")
		w.Header().Set("Cache-Control", "no-store")
		_, _ = w.Write([]byte(payload))
		return
	}

	code, err := qr.Encode(payload, cardQRLevel)
	if err != nil {
		// A credential too large to fit a QR is a real possibility as fields
		// are added, and it must not be a mystery when it happens.
		httpx.WriteError(w, http.StatusInsufficientStorage, "credential_does_not_fit_a_card",
			"this credential does not fit in a printable QR (%d characters of payload): %v",
			len(payload), err)
		return
	}

	w.Header().Set("Content-Type", "text/html; charset=utf-8")
	w.Header().Set("Cache-Control", "no-store")
	_, _ = w.Write([]byte(renderCard(c, payload, code.PNG(), h.d.Clock.Now())))
}

// renderCard builds the printable page.
//
// Deliberately one self-contained document with the image inlined: the place
// this gets printed may have a printer and no internet, and a card whose QR is
// an <img src> to a server is a card that prints as a broken image icon in
// exactly the deployment that needs it most.
func renderCard(c issuedCredential, payload string, png []byte, now time.Time) string {
	var doc map[string]any
	_ = json.Unmarshal(c.Doc, &doc)
	subject, _ := doc["credentialSubject"].(map[string]any)

	field := func(k string) string {
		if subject == nil {
			return ""
		}
		v, ok := subject[k]
		if !ok {
			return ""
		}
		return fmt.Sprint(v)
	}
	outcome := ""
	if o, ok := subject["outcome"].(map[string]any); ok {
		outcome = fmt.Sprintf("%v %v", o["value"], o["unit"])
	}
	period := ""
	if p, ok := subject["period"].(map[string]any); ok {
		period = fmt.Sprint(p["start"])
		if end, ok := p["end"]; ok {
			period += " – " + fmt.Sprint(end)
		}
	}

	rows := [][2]string{
		{"work", field("activity")},
		{"amount", outcome},
		{"when", period},
		{"issued by", field("issuerOrg")},
	}
	var dl strings.Builder
	for _, row := range rows {
		if row[1] == "" {
			continue
		}
		fmt.Fprintf(&dl, "<dt>%s</dt><dd>%s</dd>", html.EscapeString(row[0]), html.EscapeString(row[1]))
	}

	return fmt.Sprintf(`<!doctype html><meta charset="utf-8">
<title>Work record</title>
<style>
 @page { margin: 10mm }
 body { font: 13px/1.5 system-ui, sans-serif; margin: 0; display: flex; justify-content: center; padding: 16px }
 .card { width: 86mm; border: 1px solid #222; border-radius: 3mm; padding: 6mm }
 h1 { font-size: 14px; margin: 0 0 1mm }
 .sub { color: #555; font-size: 11px; margin: 0 0 4mm }
 img { width: 100%%; image-rendering: pixelated; border: 1px solid #eee }
 dl { display: grid; grid-template-columns: auto 1fr; gap: 1mm 4mm; font-size: 11px; margin: 4mm 0 0 }
 dt { color: #555 } dd { margin: 0 }
 .foot { margin-top: 4mm; font-size: 9px; color: #555; border-top: 1px solid #eee; padding-top: 2mm }
 @media print { body { padding: 0 } .card { border: none } }
</style>
<div class="card">
  <h1>%s</h1>
  <p class="sub">%s</p>
  <img alt="Scan to read this record" src="data:image/png;base64,%s">
  <dl>%s</dl>
  <p class="foot">Scanning this code reads the whole signed record, with no internet needed.
  It carries no score and no identity number: how much weight it deserves is worked out
  by whoever checks it, from what is inside. Printed %s.</p>
</div>`,
		html.EscapeString(orDefault(field("activity"), "Work record")),
		html.EscapeString(field("issuerOrg")),
		base64.StdEncoding.EncodeToString(png),
		dl.String(),
		now.UTC().Format("2 January 2006"))
}

func orDefault(s, def string) string {
	if s == "" {
		return def
	}
	return s
}
